### PR TITLE
Add JSON as a format option for `targets` command in `build_examples.py`

### PR DESCRIPTION
**Pull request by nekleo**
The `targets` command outputs the targets that are available for the `build` and `gen` commands in a string format that is easy to read but hard to parse. This change adds a option for the `targets` command to output JSON so that the available targets can be more easily parsed. To do this, ToDict methods were added to BuildTarget and TargetPart classes. When the `--format json` option is used with the `targets` command `build_examples.py` calls the ToDict method of each BuildTarget object and outputs to stdout a list of Dicts in JSON, one for each BuildTarget.

### CODE REVIEW

File: scripts/build/build/target.py [hunk 1]
```diff
@@ -95,6 +95,24 @@ def Accept(self, full_input: str):
 
         return True
 
+    def ToDict(self):
+        """Converts a TargetPart into a dictionary
+        """
+
+        result: Dict[str, str] = {}
+        result['name'] = self.name
+        bld_args_keys = list(self.build_arguments.keys())
```
Comment (andy31415) [inline scripts/build/build/target.py:104]: build_arguments seems to already be a dict, so how about:

```python
for key, value in self.build_arguments.items():
    result[key] = str(value)
```

I also wonder if it would be maybe safer to put this in `result['build_arguments']` since there seems no guarantee that build_arguments does not contain something named `name` or `only_if_re` or such.

File: scripts/build/build/target.py [hunk 2]
```diff
@@ -264,6 +282,50 @@ def HumanString(self):
...
+        result: Dict[str, Any] = {}
+
+        result['name'] = self.name
+        result['builder'] = str(self.builder_class)
+        result['shorthand'] = self.HumanString()
+        result['parts'] = []
+        result['modifiers'] = []
+
+        for target in self.fixed_targets:
+            for target_part in target:
```
Comment (andy31415) [inline scripts/build/build/target.py:321]: maybe shorter here and below:

```python
result['parts'] = [part.toDict() for part in target]
```

File: scripts/build/build_examples.py [hunk 3]
```diff
@@ -177,21 +178,34 @@ def cmd_generate(context):
...
+    if format == 'expanded':
+        for target in build.targets.BUILD_TARGETS:
             build.target.report_rejected_parts = False
             for s in target.AllVariants():
                 print(s)
-        else:
+    elif format == 'json':
+        targets = []
+
+        for target in build.targets.BUILD_TARGETS:
```
Comment (andy31415) [inline scripts/build/build_examples.py:203]: targets = [target.ToDict() for target in build.targets.BUILD_TARGETS]

File: scripts/build/build/target.py (hunk 1 above)
Comment (nekleo) [reply scripts/build/build/target.py:104]: Oops. This code is from when I was processing the TargetPart obects from BuildTarget and needed to get access to the underlying data.

File: scripts/build/build/target.py [hunk 4]
```diff
@@ -264,6 +282,50 @@ def HumanString(self):
...
         return result
 
+    def ToDict(self):
+        """Outputs a parseable description of the available variants
+        and modifiers:
+
+            like:
+
+            "foo": {
+                "builder": "<class 'builders.foo.FooBuilder'>",
```
[the remark this answers is not in the document]
Comment (nekleo) [reply scripts/build/build/target.py:292]: I don't think we need it. I'll take it out.

File: scripts/build/build/target.py (hunk 1 above)
Comment (nekleo) [reply scripts/build/build/target.py:104]: Yes, makes sense, that's safer to put in it build_arguments.

File: scripts/build/build/target.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nekleo) [reply scripts/build/build/target.py:292]: This change started to feel like object serialization, so I thought maybe I should be complete. That was my rationale but it's not needed for understanding the available targets.

File: scripts/build/build/target.py [hunk 5]
```diff
@@ -264,6 +282,50 @@ def HumanString(self):
...
+                        "name": "modifier1",
+                        "m1": "True"
+                    }
+                ]
+            },
+        """
+        result: Dict[str, Any] = {}
+
+        result['name'] = self.name
+        result['builder'] = str(self.builder_class)
```
[the remark this answers is not in the document]
Comment (nekleo) [reply scripts/build/build/target.py:315]: I'll delete it

File: scripts/build/build/target.py [hunk 6]
```diff
@@ -264,6 +285,50 @@ def HumanString(self):
...
 
+    def ToDict(self):
+        """Outputs a parseable description of the available variants
+        and modifiers:
+
+            like:
+
+            {
+                "name": "foo"
+                "shorthand": "foo-bar-baz[-m1]
```
Comment (arkq) [inline scripts/build/build/target.py:296]:
```suggestion
                "shorthand": "foo-bar-baz[-m1]"
```

File: scripts/build/build/target.py [hunk 7]
```diff
@@ -264,6 +285,50 @@ def HumanString(self):
...
+            }
+        """
+        result: Dict[str, Any] = {}
+
+        result['name'] = self.name
+        result['shorthand'] = self.HumanString()
+
+        result['parts']: List[str] = []
+        for target in self.fixed_targets:
+            result['parts'] = [part.ToDict() for part in target]
```
Comment (arkq) [inline scripts/build/build/target.py:326]: Aren't you overwriting the same dict key over and over again?

Also, whey the dict is created key-by-key instead of e.g.:
```python
return {
 'name': self.name,
 'shorthand': self.HumanString(),
  # I'm not sure whether such output is intended...
 'parts': [part.ToDict() for target in self.fixed_targets for part in target], 
 'modifiers': [part.ToDict() for part in self.modifiers],
}
```

Comment (nekleo) [reply scripts/build/build/target.py:326]: Yes, it is. Since most parts only had 1 item, the output looked ok. Parts like this were incorrect:
`TargetPart('BL602-IoT-Matter-V1', board=BouffalolabBoard.BL602_IoT_Matter_V1, module_type="BL602"),`

File: scripts/build/build/target.py [hunk 8]
```diff
@@ -264,6 +285,44 @@ def HumanString(self):
...
+                        "name": "modifier1",
+                        "m1": "True"
+                    }
+                ]
+            }
+        """
+        return {
+            'name': self.name,
+            'shorthand': self.HumanString(),
+            'parts': [part.ToDict() for target in self.fixed_targets for part in target],
```
Comment (andy31415) [inline scripts/build/build/target.py:322]: we may want an extra level of nesting here since this flattens parts. like `[[part.toDict() for part in target] for target in self.fixed_targets]`

**Review by andy31415 — APPROVED** [review]
